flang/lib/Optimizer/Builder/FIRBuilder.cpp

```diff
@@ -721,6 +721,69 @@ fir::factory::getNonDeferredLengthParams(const fir::ExtendedValue &exv) {
       [&](const auto &) -> llvm::SmallVector<mlir::Value> { return {}; });
 }
 
+// If valTy is a box type, then we need to extract the type parameters from
+// the box value.
+static llvm::SmallVector<mlir::Value> getFromBox(mlir::Location loc,
+                                                 fir::FirOpBuilder &builder,
+                                                 mlir::Type valTy,
+                                                 mlir::Value boxVal) {
+  if (auto boxTy = valTy.dyn_cast<fir::BoxType>()) {
+    auto eleTy = fir::unwrapAllRefAndSeqType(boxTy.getEleTy());
+    if (auto recTy = eleTy.dyn_cast<fir::RecordType>()) {
+      if (recTy.getNumLenParams() > 0) {
+        // Walk each type parameter in the record and get the value.
+        TODO(loc, "generate code to get LEN type parameters");
+      }
+    } else if (auto charTy = eleTy.dyn_cast<fir::CharacterType>()) {
+      if (charTy.hasDynamicLen()) {
+        auto idxTy = builder.getIndexType();
+        auto eleSz = builder.create<fir::BoxEleSizeOp>(loc, idxTy, boxVal);
+        auto kindBytes =
+            builder.getKindMap().getCharacterBitsize(charTy.getFKind()) / 8;
+        mlir::Value charSz =
+            builder.createIntegerConstant(loc, idxTy, kindBytes);
+        mlir::Value len =
+            builder.create<mlir::arith::DivSIOp>(loc, eleSz, charSz);
+        return {len};
+      }
+    }
+  }
+  return {};
+}
+
+llvm::SmallVector<mlir::Value>
+fir::factory::getTypeParams(mlir::Location loc, fir::FirOpBuilder &builder,
+                            const fir::ExtendedValue &exv) {
+  return exv.match(
+      [&](fir::UnboxedValue val) {
+        auto valTy = fir::unwrapRefType(val.getType());
+        if (auto boxTy = valTy.dyn_cast<fir::BoxType>())
```

Comment:
Is this case used, or is added in case this could happen ?

I feel like propagating fir.box in `UnboxedValue` is debatable, most of lowering tools do not expect it, and I think `fir::BoxValue` should be used instead.

Reply:
I'll remove this. I believe it is a vestigial holdover from an early attempt.